scipy/ndimage/src/ni_iterators.c

```diff
@@ -0,0 +1,331 @@
+        if (nditer->coords[axis] < nditer->dims_1[axis]) {
+            ++(nditer->coords[axis]);
+            nditer->data += nditer->strides[axis];
+            break;
+        }
+        else {
+            nditer->coords[axis] = 0;
+            nditer->data -= nditer->backstrides[axis];
+        }
+    }
+
+    return 1;
+}
+
+static int
+_NDI_NextForTwo(NDIterator *nditer1, NDIterator *nditer2)
+{
+    int axis;
+
+    ++(nditer1->index);
+    ++(nditer2->index);
+    if (nditer1->index >= nditer1->size) {
+        return 0;
+    }
+
+    for (axis = nditer1->ndim_1; axis >= 0; --axis) {
+        if (nditer1->coords[axis] < nditer1->dims_1[axis]) {
+            ++(nditer1->coords[axis]);
+            ++(nditer2->coords[axis]);
+            nditer1->data += nditer1->strides[axis];
+            nditer2->data += nditer2->strides[axis];
+            break;
+        }
+        else {
+            nditer1->coords[axis] = 0;
+            nditer2->coords[axis] = 0;
+            nditer1->data -= nditer1->backstrides[axis];
+            nditer2->data -= nditer2->backstrides[axis];
+        }
+    }
+
+    return 1;
+}
+
+
+void
+NDI_Delete(NDIterator *nditer)
+{
+    if (nditer != NULL) {
+        Py_DECREF(nditer->array);
+        free(nditer);
+    }
+}
+
+
+/*
+ ***********************************************************************
+ ***                       LineBufferIterator                        ***
+ ***********************************************************************
+ */
+
+struct LineBufferIterator_Internal {
+    /* Input and output buffers. */
+    npy_double *in_buffer;
+    npy_double *out_buffer;
+    /* Size of filter. */
+    npy_intp filter_size;
+    /* Size of filter before and after origin, size = before + 1 + after. */
+    npy_intp size_before;
+    npy_intp size_after;
+    /* Length of array lines. */
+    npy_intp line_length;
+    /* Iterators over input and output array lines. */
+    NDIterator *in_nditer;
+    NDIterator *out_nditer;
+    /* Line strides of the input and output arrays. */
+    npy_intp in_stride;
+    npy_intp out_stride;
+    /* Functions to read and write array lines from/to buffers. */
+    read_line_func *read_func;
+    write_line_func *write_func;
+    /* Extend function and value for constant mode. */
+    extend_line_func *extend_func;
+    npy_double extend_value;
+};
+
+
+static NPY_INLINE void
+_LBI_ExtendLine(LineBufferIterator *lbiter)
+{
+    lbiter->extend_func(lbiter->in_buffer, lbiter->line_length,
+                        lbiter->size_before, lbiter->size_after,
+                        lbiter->extend_value);
+}
+
+
+static NPY_INLINE void
+_LBI_ReadLine(LineBufferIterator *lbiter)
+{
+    lbiter->read_func(lbiter->in_buffer + lbiter->size_before,
+                      lbiter->in_nditer->data,
+                      lbiter->line_length, lbiter->in_stride);
+    _LBI_ExtendLine(lbiter);
+}
+
+
+LineBufferIterator*
+LBI_New(PyArrayObject *input, int axis, PyArrayObject *output,
```

Comment:
This one is so good it will always feel like new... ;-)
It's supposed to be `new` as in `new` and `delete` in C++, i.e. the "class" constructor. It's inspired by the numpy iterator's [`NpyIter_New`](https://docs.scipy.org/doc/numpy/reference/c-api.iterator.html#c.NpyIter_New).